log: Avoid side effects by using own logger instance - fixes #8907

srebhan opened the pull request:
#### What is the purpose of this change?

Commit dfa4d94 replaced the old logging with the more modern `slog` package. However, this change is intrusive for applications using `slog` as their default logging facility, causing side effects and a fragile construct when using `slog` logging with custom output functions.

This PR sets up a custom logger, avoiding side effects from overriding the `slog` default logger.

#### Was the change discussed in an issue or in the forum before?

resolves #8907 

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [ ] I have added tests for all changes in this PR if appropriate.
- [ ] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

srebhan committed:
log: Avoid side effects by using own logger instance - fixes #8907

ncw reviewed (commented):
Thanks for fixing this. Using rclone as a library is probably the least well supported way of using rclone but we need to play nice if people do :-)

I put some queries inline.

srebhan commented on fs/log/slog.go:
```diff
@@ -36,11 +36,8 @@ func defaultHandler() *OutputHandler {
-	slog.SetDefault(slog.New(h))
-
-	// Make log.Printf logs at level Notice
-	slog.SetLogLoggerLevel(fs.SlogLevelNotice)
```
Well this line overrides the log-bridge of slog! This should not be done within the library but rather in the application so others can use the library without their logging being changed...

ncw commented:
@srebhan there's a fundamental tension here.  Rclone was designed as a command line application. It has a API interface in the remote control interface. Both of those are well supported interfaces for rclone. linking with rclone directly and using it as a package is very much a third class citizen. The changes we've got so far in the pull request break rclone as a library and as an API server. So I think what we need to do is put a global flag in the log package `var InitialiseLogger = true` which only makes those logging changes when that flag is `false`.  So you could clear the flag in your initialisation and then rclone won't initialise its log package in the way it does at the moment. How does that sound to you?

ncw committed:
log: add InitialiseDefaultLogger flag to avoid side effects for library users

The previous commit stopped rclone from calling slog.SetDefault(),
which broke log.Print/log.Fatal capture for CLI and API server use.

Instead, keep rclone's private logger (so internal logging always
works), but gate the slog.SetDefault() and slog.SetLogLoggerLevel()
calls behind a new fs.InitialiseDefaultLogger flag (default true).

Library consumers (e.g. Telegraf) can set fs.InitialiseDefaultLogger
= false before rclone's logging is initialised to prevent rclone from
overriding the process-wide default slog logger, while rclone's own
logging continues to work correctly through its private instance.

Fixes #8907

srebhan commented:
@ncw any update? This is a minimal code-example to trigger the issue (see #8907):

```go
package main

import (
	"context"
	"log/slog"

	"github.com/rclone/rclone/fs"
	"github.com/rclone/rclone/fs/log"
)

func main() {
	// This logging line is already influenced by the import of fs/log which
	// sets the slog-handler potentially causing side effects in any application
	// using it.
	slog.Info("This is the slog logger!")

	// Setting the handler output to anything using the default slog logger
	// will cause an infinite loop as the default handler will call this
	// function which in turn will call the default handler during processing.
	log.Handler.SetOutput(func(level slog.Level, text string) {
		slog.Log(context.Background(), level, "[myprefix] "+text)
	})

	// This line will trigger the infinite loop mentioned above.
	fs.Log(nil, "A message from rclone FS...")
}
```
your change to this PR breaks the setup again.

srebhan commented:
@ncw can we please restore my original PR and get this merged? I made sure this has no negative effect on rclone itself as it is backward compatible. This PR sits there for a while and Telegraf is now seeing some vulnerabilities due to an old rclone dependency and we cannot upgrade...

ncw committed:
log: move slog.SetDefault out of package init for library users

The previous commit added an InitialiseDefaultLogger flag, but the flag
is consulted from defaultHandler() which runs at package-init time when
fs/log is imported. Library consumers (e.g. Telegraf) cannot flip the
flag before that init runs, so the side effect on slog.Default() still
occurs and the infinite loop reproduced in #8907 still happens.

Instead, split the two concerns cleanly:

- Importing fs/log only sets rclone's private logger via fs.SetLogger,
  so internal rclone logging works from the moment the package is
  imported but the process-wide slog default is left untouched.

- slog.SetDefault and slog.SetLogLoggerLevel move into InitLogging,
  which is called explicitly from the CLI (cmd/cmd.go), the librclone
  wrapper and the integration test framework. So rclone-as-a-program
  keeps capturing log.Print/log.Fatal and slog.Default() output as
  before.

Library consumers that import fs/log without calling InitLogging now
keep their own slog default and can safely route rclone output back
into it via log.Handler.SetOutput without recursing.

Fixes #8907

ncw commented:
Sorry for the delay looking at this @srebhan 

I've added another commit to the branch which restores the default logging but only if you call `InitLogging` and not on package init(). I think this should fix your concerns and the combined diff looks sensible now.

Let me know what you think.

Thanks

srebhan commented:
@ncw no worries. Your fix works for me! Can we please merge this and get a release so we can fix the CVE in Telegraf?

ncw reviewed (approved):
Thanks for reviewing @srebhan - will merge now.

srebhan commented:
@ncw do you have any ETA for a release containing this commit?

ncw commented:
It is scheduled for v1.75 which you can find the release date in the milestone.

I didn't want to put it in the point release as I deemed it a bit too risky. Alas it just missed the v1.74 release.
